### PR TITLE
a rough idea about the Rust API

### PR DESCRIPTION
**Pull request by lum1n0us**
The PR is trying to define the rust API layout (physically and logically). Based on those, finish examples (in comments) and documents.

### CODE REVIEW

**Review by Mossaka — COMMENTED** [review]
Hey thanks for starting this work! I am really excited for this. Let me know if you need a review when the PR is ready!

File: language-bindings/rust/src/instance.rs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    helper::error_buf_to_string, helper::DEFAULT_ERROR_BUF_SIZE, module::Module, RuntimeError,
+};
+
+#[derive(Debug)]
+pub struct Instance {
+    instance: wasm_module_inst_t,
+}
+
+impl Instance {
+    pub fn new(module: &Module, stack_size: u32) -> Result<Self, RuntimeError> {
```
Comment (wenyongh) [inline language-bindings/rust/src/instance.rs:18]: Had better add `heap_size: u32` or `host_managed_heap_size: u32` also, like wasm_runtime_instantiate, to allow developer to run the `nostdlib` wasm module. We may add some comments to help developer to pass the correct values, but had better not ignore it.

File: language-bindings/rust/src/instance.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        module: &Module,
+        stack_size: u32,
+        heap_size: u32,
+    ) -> Result<Self, RuntimeError> {
+        let mut error_buf = [0 as c_char; DEFAULT_ERROR_BUF_SIZE];
+        let instance = unsafe {
+            wasm_runtime_instantiate(
+                module.get_inner_module(),
+                stack_size,
+                0,
```
Comment (wenyongh) [inline language-bindings/rust/src/instance.rs:33]: Should be heap_size.

File: language-bindings/rust/crates/wamr-sys/build.rs [hunk 3]
```diff
@@ -10,22 +10,34 @@ use cmake::Config;
 use std::{env, path::PathBuf};
 
 fn main() {
-    let dst = Config::new("../..").build();
+    let wamr_root = "../../../../";
+    let llvm_dir = format!("{wamr_root}core/deps/llvm/build");
+    let dst = Config::new(wamr_root)
+        .define("LLVM_DIR", llvm_dir)
+        .define("WAMR_BUILD_INTERP", "1")
+        .define("WAMR_BUILD_FAST_INTERP", "0")
```
Comment (wenyongh) [inline language-bindings/rust/crates/wamr-sys/build.rs:18]: Is it better to enable fast interpreter by default?

File: language-bindings/rust/Cargo.toml [hunk 4]
```diff
@@ -1,19 +1,12 @@
-# Copyright (C) 2023 Liquid Reply GmbH. All rights reserved.
-# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
```
Comment (wenyongh) [inline language-bindings/rust/Cargo.toml:2]: Had better keep license header?

File: language-bindings/rust/src/function.rs [hunk 5]
```diff
@@ -0,0 +1,142 @@
+//! a wasm function. only exported.
```
Comment (wenyongh) [inline language-bindings/rust/src/function.rs:1]: Had better add license header?

File: language-bindings/rust/src/function.rs [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+                &mut result_type,
+            );
+        }
+
+        match result_type as u32 {
+            wasm_valkind_enum_WASM_I32 => Ok(WasmValue::decode_to_i32(result)),
+            wasm_valkind_enum_WASM_I64 => Ok(WasmValue::decode_to_i64(result)),
+            wasm_valkind_enum_WASM_F32 => Ok(WasmValue::decode_to_f32(result)),
+            wasm_valkind_enum_WASM_F64 => Ok(WasmValue::decode_to_f64(result)),
+            _ => Err(RuntimeError::InvalidResult),
```
Comment (wenyongh) [inline language-bindings/rust/src/function.rs:53]: Not sure whether void return type is handled?

File: language-bindings/rust/src/helper.rs [hunk 7]
```diff
@@ -0,0 +1,52 @@
+use std::ffi::CStr;
```
Comment (wenyongh) [inline language-bindings/rust/src/helper.rs:1]: license header?

File: language-bindings/rust/src/instance.rs [hunk 8]
```diff
@@ -0,0 +1,106 @@
+//! an instantiated module. The module is instantiated with the given imports.
```
Comment (wenyongh) [inline language-bindings/rust/src/instance.rs:1]: license header?

File: language-bindings/rust/src/value.rs [hunk 9]
```diff
@@ -0,0 +1,110 @@
+//! a wasm value. Always used as function parameters and results
```
Comment (wenyongh) [inline language-bindings/rust/src/value.rs:1]: license header?

**Comment by lum1n0us** [comment]
> Hey thanks for starting this work! I am really excited for this. Let me know if you need a review when the PR is ready!

Thanks. This one is ready for comments.

File: language-bindings/rust/src/instance.rs [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    ///
+    /// # Error
+    ///
+    /// Return `RuntimeError::CompilationError` if failed.
+    pub fn new_with_args(
+        module: &Module,
+        stack_size: u32,
+        heap_size: u32,
+    ) -> Result<Self, RuntimeError> {
+        let init_thd_env = unsafe { wasm_runtime_init_thread_env() };
```
Comment (wenyongh) [inline language-bindings/rust/src/instance.rs:49]: Should call `wasm_runtime_destroy_thread_env` somewhere?

Comment (lum1n0us) [reply language-bindings/rust/src/instance.rs:49]: nice catch. it is a good one,

**Comment by lum1n0us** [comment]
@0xE282B0 @Mossaka Please let us know your opinions about this PR. thanks a lot.

File: language-bindings/rust/resources/test/gcd/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,16 @@
+#[no_mangle]
```
Comment (wenyongh) [inline language-bindings/rust/resources/test/gcd/src/main.rs:1]: Should we add license header for files under this folder?

File: language-bindings/rust/README.md [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+Say there is a *test.wasm* includes a function named *add*.
+
+``` wat
+  (func (export "add") (param i32 i32) (result i32)
+    (local.get 0)
+    (local.get 1)
+    (i32.add)
+  )
+```rust
+The rust code to call the *add* function likes below:
```
[the remark this answers is not in the document]
Comment (lum1n0us) [reply language-bindings/rust/README.md:67]: may lead to two verbs. *is* and *like*.

File: language-bindings/rust/README.md [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+  let params: Vec<WasmValue> = vec![WasmValue::I32(1), WasmValue::I32(2)];
+  let ret = wasm_func.call(&instance, &params)?;
+  assert_eq!(ret , WasmValue::I32(13));
+
+  Ok()
+}
+```
+
+
+License: Apache-2.0 WITH LLVM-exception
```
[the remark this answers is not in the document]
Comment (lum1n0us) [reply language-bindings/rust/README.md:141]: *README.md* is a generated by cargo. If not bother too much, I'd like to keep it.

[the remark this answers is not in the document]
Comment (wenyongh) [reply language-bindings/rust/README.md:141]: OK, thanks.

File: language-bindings/rust/README.md [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+
+Say there is a *test.wasm* includes a function named *add*.
+
+``` wat
+  (func (export "add") (param i32 i32) (result i32)
+    (local.get 0)
+    (local.get 1)
+    (i32.add)
+  )
+```rust
```
Comment (wenyongh) [inline language-bindings/rust/README.md:66]: Should be ````?

File: language-bindings/rust/README.md [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+  (func $extra (import "extra") (result i32))
+  (func (export "add") (param i32 i32) (result i32)
+    (local.get 0)
+    (local.get 1)
+    (i32.add)
+    (call $extra)
+    (i32.add)
+  )
+)
+```rust
```
Comment (wenyongh) [inline language-bindings/rust/README.md:112]: Should be ```?

File: language-bindings/rust/README.md [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+- *Runtime*. It is the environment that hosts all the wasm modules. Each process has one runtime instance.
+- *Module*. It is the compiled .wasm or .aot. It can be loaded into runtime and instantiated into instance.
+- *Instance*. It is the running instance of a module. It can be used to call export functions.
+- *Function*. It is the exported function.
+
+#### WASI concepts
+
+- *WASIArgs*. It is used to configure the WASI environment.
+  - *pre-open*. All files and directories in the list will be opened before the .wasm or .aot loaded.
+  - *allowed address*. all ip addresses in the *allowed address* list will be allowed to connect with a socket.
```
Comment (wenyongh) [inline language-bindings/rust/README.md:34]: Had better `All ip addresses`

**Review by wenyongh — COMMENTED** [review]
LGTM with several minor comments.

**Pull request merged**